src/components/SigningModals/TransactionStepper/TransactionStepper.js

```diff
@@ -0,0 +1,247 @@
+  if (action === 'setHash') {
+    steps[stepIndex].hash = value
+    return [...steps]
+  }
+  if (action === 'setStatus') {
+    steps[stepIndex].status = value
+    return [...steps]
+  }
+  return steps
+}
+
+function TransactionStepper({ steps, onComplete, className }) {
+  const theme = useTheme()
+  const [stepperStage, setStepperStage] = useState(0)
+  const [stepState, updateStep] = useReducer(
+    reduceSteps,
+    initialStepState(steps)
+  )
+  const [firstStart, setFirstStart] = useState(true)
+
+  const {
+    outerBoundsRef,
+    innerBoundsRef,
+    layout,
+    measuring,
+  } = useStepperLayout()
+
+  const canPerformUpdate = useRef(true)
+
+  const stepsCount = steps.length - 1
+
+  const onStart = useCallback(() => {
+    // Don’t animate on first render
+    if (firstStart) {
+      setFirstStart(false)
+    }
+  }, [firstStart])
+
+  const renderStep = (stepIndex, showDivider) => {
+    const title = steps[stepIndex][0]
+    const { status, hash } = stepState[stepIndex]
+
+    return (
+      <li
+        key={stepIndex}
+        css={`
+          display: flex;
+        `}
+      >
+        <Step
+          title={title}
+          number={stepIndex + 1}
+          status={status}
+          showDivider={showDivider}
+          transactionHash={hash}
+        />
+      </li>
+    )
+  }
+
+  const renderSteps = () => {
+    return steps.map((_, index) => {
+      const showDivider = index < stepsCount
+
+      return renderStep(index, showDivider)
+    })
+  }
+
+  const updateStepStatus = useCallback(
+    status => {
+      if (canPerformUpdate.current) {
+        updateStep(['setStatus', stepperStage, status])
+      }
+    },
+    [stepperStage]
+  )
+
+  const updateHash = useCallback(
+    hash => {
+      if (canPerformUpdate.current) {
+        updateStep(['setHash', stepperStage, hash])
+      }
+    },
+    [stepperStage]
+  )
+
+  const handleSign = useCallback(() => {
+    const signProcess = steps[stepperStage][1]
+
+    // Always start new step in the "Prompting" state
+    updateStepStatus(STEP_PROMPTING)
+
+    // Pass state updates as render props to signProcess
+    signProcess({
+      setStepHash: hash => updateHash(hash),
+      setStepWorking: () => updateStepStatus(STEP_WORKING),
+      setStepError: () => updateStepStatus(STEP_ERROR),
+      setStepSuccess: () => {
+        updateStepStatus(STEP_SUCCESS)
+
+        // Advance to next step or fire complete callback
+        if (stepperStage !== stepsCount && canPerformUpdate.current) {
+          setStepperStage(stepperStage + 1)
+        }
+
+        if (stepperStage === stepsCount && canPerformUpdate.current) {
+          onComplete()
+        }
+      },
+    })
+  }, [
+    steps,
+    stepperStage,
+    updateStepStatus,
+    updateHash,
+    onComplete,
+    stepsCount,
+  ])
+
+  useEffect(handleSign, [stepperStage])
+
+  // Prevent async state updates after unmount
+  useEffect(() => {
```

Comment:
What do you think about encapsulating this into its own hook, e.g. `useDetectUnmounted()` or etc?